### PR TITLE
feat: improve Reth bindings and fix `debug_traceCallMany` and `trace_callMany` methods

### PR DESCRIPTION
**Pull request by zerosnacks**
## Motivation

Closes: https://github.com/alloy-rs/alloy/issues/1074 and improves on Reth bindings added in https://github.com/alloy-rs/alloy/pull/1092

### Fixes

- [x] Fix `debug_traceCallMany`
- [x] Fix `trace_callMany`

### Other

- `keys()` / `address()` method for Reth binding - used in tracing but useful in general so you can perform transactions
- adds `random_instance()` to reduce flakiness of concurrent spawned Reth nodes
- Adds `run_with_tempdir_*` methods for proper handling and clean up when using tempdirs + updating tests
- Unification Anvil errors with other nodes

Does not include the addition of `eth_callMany`, to be added in https://github.com/alloy-rs/alloy/issues/1274 and requires more consideration. Ties into #328 / #1119.

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

**Comment by yash-atreya** [comment]
With #1062 merged this can be picked up again @zerosnacks. Let me know if you have bandwidth for this.

**Comment by zerosnacks** [comment]
> With #1062 merged this can be picked up again @zerosnacks. Let me know if you have bandwidth for this.

Yes, I'm picking it back up

File: crates/provider/src/ext/trace.rs [hunk 1]
```diff
@@ -178,18 +178,161 @@ where
...
+            //         },
+            //         [Trace],
+            //     ),
+            // ]
+
+            // Err(ErrorResp(ErrorPayload { code: -32602, message: "Invalid params", data:
+            // Some(RawValue("invalid type: map, expected a tuple of size 2 at line 1 column 1"))
+            // }))
+        })
+        .await
```
Comment (zerosnacks) [inline crates/provider/src/ext/trace.rs:336]: @yash-atreya I'm stuck on a types / serialization issue here - mind having a look?

in Reth:

The signature:

```rust
    async fn trace_call_many(
        &self,
        calls: Vec<(TransactionRequest, HashSet<TraceType>)>,
        block_id: Option<BlockId>,
    ) -> RpcResult<Vec<TraceResults>>;
```

```rust
#[tokio::test(flavor = "multi_thread")]
async fn test_trace_call_many() {
    let handle = launch_http(vec![RethRpcModule::Trace]).await;
    let client = handle.http_client().unwrap();

    let x = TraceApiClient::trace_call_many(
        &client,
        vec![
            (TransactionRequest::default(), [TraceType::Trace].into()),
            (TransactionRequest::default(), [TraceType::Trace].into()),
        ],
        Some(BlockNumberOrTag::Latest.into()),
    )
    .await
    .unwrap();

    println!("{:?}", x);
}
```

yields:

```
[TraceResults { output: 0x, state_diff: None, trace: [TransactionTrace { action: Create(CreateAction { from: 0x0000000000000000000000000000000000000000, gas: 49947000, init: 0x, value: 0 }), error: None, result: Some(Create(CreateOutput { address: 0xbd770416a3345f91e4b34576cb804a576fa48eb1, code: 0x, gas_used: 0 })), subtraces: 0, trace_address: [] }], vm_trace: None }, TraceResults { output: 0x, state_diff: None, trace: [TransactionTrace { action: Create(CreateAction { from: 0x0000000000000000000000000000000000000000, gas: 49947000, init: 0x, value: 0 }), error: None, result: Some(Create(CreateOutput { address: 0x5a443704dd4b594b382c22a083e2bd3090a6fef3, code: 0x, gas_used: 0 })), subtraces: 0, trace_address: [] }], vm_trace: None }]
```

Confirming that the internals do work.

Full error is of this format:

```sh
2024-09-10T11:08:27.380824Z DEBUG jsonrpsee_core::proc_macros_support: Error parsing "calls" as "Vec < (TransactionRequest, HashSet < TraceType >) >": ErrorObject { code: InvalidParams, message: "Invalid params", data: Some(RawValue("invalid type: map, expected a tuple of size 2 at line 1 column 1")) }
```

Comment (zerosnacks) [reply crates/provider/src/ext/trace.rs]: Turns out it was related to the formatting when passed to the `request` method not being a tuple (`(request,))` vs `request`)

File: crates/provider/src/ext/trace.rs [hunk 2]
```diff
@@ -10,9 +10,6 @@ use alloy_rpc_types_trace::{
 };
 use alloy_transport::{Transport, TransportResult};
 
-/// List of trace calls for use with [`TraceApi::trace_call_many`]
-pub type TraceCallList<'a, 'b, N> = [(&'a <N as Network>::TransactionRequest, &'b [TraceType])];
```
Comment (mattsse) [inline crates/provider/src/ext/trace.rs]: why remove this alias?

File: crates/rpc-types-eth/src/call.rs [hunk 3]
```diff
@@ -15,6 +15,7 @@ pub struct Bundle {
     /// All transactions to execute
     pub transactions: Vec<TransactionRequest>,
     /// Block overrides to apply
+    #[serde(skip_serializing_if = "Option::is_none")]
```
Comment (mattsse) [inline crates/rpc-types-eth/src/call.rs:18]:
```suggestion
    #[serde(default, skip_serializing_if = "Option::is_none")]
```

**Review by mattsse — CHANGES_REQUESTED** [review]
please undo all the unrelated changes here

File: crates/provider/src/ext/trace.rs (hunk 2 above)
Comment (zerosnacks) [reply crates/provider/src/ext/trace.rs]: Don't really see a need for it but adding it back to prevent adding unnecessary breaking changes

**Comment by zerosnacks** [comment]
Undone unrelated changes.

Kept:
- `keys()` / `address()` method for Reth binding - used in tracing but useful in general
- adds `random_instance()` to reduce flakiness
- `run_with_tempdir_*` methods for proper handling and clean up when using tempdirs + updating tests
- Anvil unification changes

As I think those a relevant to the newly added tests.

**Review by mattsse — COMMENTED** [review]
could we please split this into two, the changes to node-bindings and provider are unrelated, right?

**Comment by zerosnacks** [comment]
Makes sense, agree that is better. 

Split into two PRs: https://github.com/alloy-rs/alloy/pull/1278, https://github.com/alloy-rs/alloy/pull/1279

**Pull request closed without merging**